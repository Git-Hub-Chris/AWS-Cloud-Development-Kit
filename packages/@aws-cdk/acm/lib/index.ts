export * from './certificate';
export * from './certificate-ref';

// The L1 library for AWS::CertificateManager:
export * from '../cfn/certificatemanager';
