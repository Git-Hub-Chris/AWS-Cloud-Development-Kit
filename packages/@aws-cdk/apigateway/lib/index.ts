// The L1 Library for AWS::ApiGateway:
export * from '../cfn/apigateway';
