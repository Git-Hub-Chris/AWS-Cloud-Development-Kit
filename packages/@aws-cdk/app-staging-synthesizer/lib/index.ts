export * from './default-staging-stack';
export * from './synthesizer';
