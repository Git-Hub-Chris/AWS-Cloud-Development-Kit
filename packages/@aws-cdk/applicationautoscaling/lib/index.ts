// The L1 Library for AWS::ApplicationAutoScaling:
export * from '../cfn/applicationautoscaling';
