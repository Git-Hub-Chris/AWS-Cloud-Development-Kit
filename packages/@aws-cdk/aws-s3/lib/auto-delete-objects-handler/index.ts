// eslint-disable-next-line import/no-extraneous-dependencies
import { S3, CloudFormation } from 'aws-sdk';
import * as yaml from 'yaml';

/**
 * The custom resource should clear the bucket in one of the following cases:
 *
 * - The stack is deleted
 * - The target bucket is removed from the template
 * - The target bucket is replaced
 * - The target bucket is created in a deployment that gets rolled back
 *
 * In particular, it should NOT clear the bucket in a case the custom resource is deleted
 * but the target bucket is unaffected. This could happen in the following cases:
 *
 * - The autoDelete feature used to be turned on, and now gets turned off (leads to removal
 *   of the CR from the template, without affecting the bucket)
 * - The autoDelete feature used to be turned off, now gets turned on, but the deployment
 *   gets rolled back (leads to creation and immediate deletion of the CR, without
 *   affecting the bucket).
 *
 * The only cases where we might misclassify is when the CR gets deleted. To
 * determine whether or not we should empty the bucket, during a `Delete` event
 * we will look at the stack state and depending on the state of the stack
 * (rolling forward or rolling backward), compare the OLD and NEW templates to
 * determine whether the bucket should be present in the final state:
 *
 * - ROLL FORWARD: delete contents if the bucket is not in the NEW template
 * - ROLLBACK: delete contents if the bucket is not in the OLD template
 */

const s3 = new S3();
const cfn = new CloudFormation();

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
  switch (event.RequestType) {
    case 'Create':
      return;
    case 'Update':
      return onUpdate(event);
    case 'Delete':
      return onDelete(event.StackId, event.ResourceProperties?.BucketLogicalId, event.ResourceProperties?.BucketName);
  }
}

async function onUpdate(event: AWSLambda.CloudFormationCustomResourceEvent) {
  const updateEvent = event as AWSLambda.CloudFormationCustomResourceUpdateEvent;
  const oldBucketName = updateEvent.OldResourceProperties?.BucketName;
  const newBucketName = updateEvent.ResourceProperties?.BucketName;
  const bucketNameHasChanged = newBucketName != null && oldBucketName != null && newBucketName !== oldBucketName;

  /* If the name of the bucket has changed, CloudFormation will try to delete the bucket
     and create a new one with the new name. So we have to delete the contents of the
     bucket so that this operation does not fail. */
  if (bucketNameHasChanged) {
    return emptyBucket(oldBucketName);
  }
}

/**
 * Recursively delete all items in the bucket
 *
 * @param bucketName the bucket name
 */
async function emptyBucket(bucketName: string) {
  const listedObjects = await s3.listObjectVersions({ Bucket: bucketName }).promise();
  const contents = [...listedObjects.Versions ?? [], ...listedObjects.DeleteMarkers ?? []];
  if (contents.length === 0) {
    return;
  }

  const records = contents.map((record: any) => ({ Key: record.Key, VersionId: record.VersionId }));
  await s3.deleteObjects({ Bucket: bucketName, Delete: { Objects: records } }).promise();

  if (listedObjects?.IsTruncated) {
    await emptyBucket(bucketName);
  }
}

async function onDelete(stackId: string, logicalId?: string, bucketName?: string) {
  if (!bucketName) {
    throw new Error('No BucketName was provided.');
  }
  if (!logicalId) {
    throw new Error('No Logical ID was provided.');
  }
  if (await isBucketAboutToBeDeleted(stackId, logicalId)) {
    await emptyBucket(bucketName);
  }
}

/**
 * Go and inspect CloudFormation to see if the target bucket is about to be deleted
 */
async function isBucketAboutToBeDeleted(stackId: string, logicalId: string) {
  const stackResponse = await cfn.describeStacks({ StackName: stackId }).promise();
  if (!stackResponse.Stacks?.[0]) {
    throw new Error(`Could not find stack with ID: ${stackId}`);
  }
  const stackStatus = stackResponse.Stacks[0].StackStatus;
  process.stdout.write(`Stack status: ${stackStatus}\n`);

  // Case 1: the stack failed creation.
  // Case 2: the stack is being deleted.
  // In both cases, by definition the bucket will go bye-bye.
  if (stackStatus === 'ROLLBACK_IN_PROGRESS' || stackStatus === 'DELETE_IN_PROGRESS') {
    return true;
  }

  // Case 3: we're cleaning up after a successful rollforward.
  // Case 4: we're rolling back a failed update.
  // In both cases, either the bucket is also being deleted here, or it's just
  // the CR that's being deleted.
  // `GetTemplate` will show us the template we are moving to ('new' in case 3,
  // 'old' in case 4). We will check if the bucket is in the template returned
  // by `GetTemplate` to see if we need to clean it.
  const destinationTemplateResponse = await cfn.getTemplate({ StackName: stackId, TemplateStage: 'Processed' }).promise();
  let template;
  try {
    template = yaml.parse(destinationTemplateResponse.TemplateBody ?? '{}', {
      schema: 'core',
    });
  } catch (e) {
    throw new Error(`Unable to parse CloudFormation template (is it not YAML?): ${destinationTemplateResponse.TemplateBody}`);
  }

  if (logicalId in (template.Resources ?? {})) {
    process.stdout.write(`Bucket ${logicalId} is in target template, so NOT cleaning.\n`);
    return false;
  } else {
    process.stdout.write(`Bucket ${logicalId} is NOT in target template, so cleaning.\n`);
    return true;
  }
}