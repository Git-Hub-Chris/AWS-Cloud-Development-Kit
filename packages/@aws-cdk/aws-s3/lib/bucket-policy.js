"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BucketPolicy = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_iam_1 = require("@aws-cdk/aws-iam");
const core_1 = require("@aws-cdk/core");
const cfn_reference_1 = require("@aws-cdk/core/lib/private/cfn-reference");
const bucket_1 = require("./bucket");
const s3_generated_1 = require("./s3.generated");
/**
 * The bucket policy for an Amazon S3 bucket
 *
 * Policies define the operations that are allowed on this resource.
 *
 * You almost never need to define this construct directly.
 *
 * All AWS resources that support resource policies have a method called
 * `addToResourcePolicy()`, which will automatically create a new resource
 * policy if one doesn't exist yet, otherwise it will add to the existing
 * policy.
 *
 * Prefer to use `addToResourcePolicy()` instead.
 */
class BucketPolicy extends core_1.Resource {
    constructor(scope, id, props) {
        super(scope, id);
        /**
         * A policy document containing permissions to add to the specified bucket.
         * For more information, see Access Policy Language Overview in the Amazon
         * Simple Storage Service Developer Guide.
         */
        this.document = new aws_iam_1.PolicyDocument();
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_s3_BucketPolicyProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, BucketPolicy);
            }
            throw error;
        }
        this.bucket = props.bucket;
        this.resource = new s3_generated_1.CfnBucketPolicy(this, 'Resource', {
            bucket: this.bucket.bucketName,
            policyDocument: this.document,
        });
        if (props.removalPolicy) {
            this.resource.applyRemovalPolicy(props.removalPolicy);
        }
    }
    /**
     * Create a mutable `BucketPolicy` from a `CfnBucketPolicy`.
     */
    static fromCfnBucketPolicy(cfnBucketPolicy) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_s3_CfnBucketPolicy(cfnBucketPolicy);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.fromCfnBucketPolicy);
            }
            throw error;
        }
        // use a "weird" id that has a higher chance of being unique
        const id = '@FromCfnBucketPolicy';
        // if fromCfnBucketPolicy() was already called on this CfnBucketPolicy,
        // return the same L2
        // (as different L2s would conflict, because of the mutation of the document property of the L1 below)
        const existing = cfnBucketPolicy.node.tryFindChild(id);
        if (existing) {
            return existing;
        }
        // resolve the Bucket this Policy references
        let bucket;
        if (core_1.Token.isUnresolved(cfnBucketPolicy.bucket)) {
            const bucketIResolvable = core_1.Tokenization.reverse(cfnBucketPolicy.bucket);
            if (bucketIResolvable instanceof cfn_reference_1.CfnReference) {
                const cfnElement = bucketIResolvable.target;
                if (cfnElement instanceof s3_generated_1.CfnBucket) {
                    bucket = bucket_1.Bucket.fromCfnBucket(cfnElement);
                }
            }
        }
        if (!bucket) {
            bucket = bucket_1.Bucket.fromBucketName(cfnBucketPolicy, '@FromCfnBucket', cfnBucketPolicy.bucket);
        }
        const ret = new class extends BucketPolicy {
            constructor() {
                super(...arguments);
                this.document = aws_iam_1.PolicyDocument.fromJson(cfnBucketPolicy.policyDocument);
            }
        }(cfnBucketPolicy, id, {
            bucket,
        });
        // mark the Bucket as having this Policy
        bucket.policy = ret;
        return ret;
    }
    /**
     * Sets the removal policy for the BucketPolicy.
     * @param removalPolicy the RemovalPolicy to set.
     */
    applyRemovalPolicy(removalPolicy) {
        this.resource.applyRemovalPolicy(removalPolicy);
    }
}
exports.BucketPolicy = BucketPolicy;
_a = JSII_RTTI_SYMBOL_1;
BucketPolicy[_a] = { fqn: "@aws-cdk/aws-s3.BucketPolicy", version: "0.0.0" };
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiYnVja2V0LXBvbGljeS5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbImJ1Y2tldC1wb2xpY3kudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6Ijs7Ozs7O0FBQUEsOENBQWtEO0FBQ2xELHdDQUE2RTtBQUM3RSwyRUFBdUU7QUFFdkUscUNBQTJDO0FBQzNDLGlEQUE0RDtBQWdCNUQ7Ozs7Ozs7Ozs7Ozs7R0FhRztBQUNILE1BQWEsWUFBYSxTQUFRLGVBQVE7SUFxRHhDLFlBQVksS0FBZ0IsRUFBRSxFQUFVLEVBQUUsS0FBd0I7UUFDaEUsS0FBSyxDQUFDLEtBQUssRUFBRSxFQUFFLENBQUMsQ0FBQztRQWJuQjs7OztXQUlHO1FBQ2EsYUFBUSxHQUFHLElBQUksd0JBQWMsRUFBRSxDQUFDOzs7Ozs7K0NBOUNyQyxZQUFZOzs7O1FBd0RyQixJQUFJLENBQUMsTUFBTSxHQUFHLEtBQUssQ0FBQyxNQUFNLENBQUM7UUFFM0IsSUFBSSxDQUFDLFFBQVEsR0FBRyxJQUFJLDhCQUFlLENBQUMsSUFBSSxFQUFFLFVBQVUsRUFBRTtZQUNwRCxNQUFNLEVBQUUsSUFBSSxDQUFDLE1BQU0sQ0FBQyxVQUFVO1lBQzlCLGNBQWMsRUFBRSxJQUFJLENBQUMsUUFBUTtTQUM5QixDQUFDLENBQUM7UUFFSCxJQUFJLEtBQUssQ0FBQyxhQUFhLEVBQUU7WUFDdkIsSUFBSSxDQUFDLFFBQVEsQ0FBQyxrQkFBa0IsQ0FBQyxLQUFLLENBQUMsYUFBYSxDQUFDLENBQUM7U0FDdkQ7S0FDRjtJQWpFRDs7T0FFRztJQUNJLE1BQU0sQ0FBQyxtQkFBbUIsQ0FBQyxlQUFnQzs7Ozs7Ozs7OztRQUNoRSw0REFBNEQ7UUFDNUQsTUFBTSxFQUFFLEdBQUcsc0JBQXNCLENBQUM7UUFFbEMsdUVBQXVFO1FBQ3ZFLHFCQUFxQjtRQUNyQixzR0FBc0c7UUFDdEcsTUFBTSxRQUFRLEdBQUcsZUFBZSxDQUFDLElBQUksQ0FBQyxZQUFZLENBQUMsRUFBRSxDQUFDLENBQUM7UUFDdkQsSUFBSSxRQUFRLEVBQUU7WUFDWixPQUFxQixRQUFRLENBQUM7U0FDL0I7UUFFRCw0Q0FBNEM7UUFDNUMsSUFBSSxNQUEyQixDQUFDO1FBQ2hDLElBQUksWUFBSyxDQUFDLFlBQVksQ0FBQyxlQUFlLENBQUMsTUFBTSxDQUFDLEVBQUU7WUFDOUMsTUFBTSxpQkFBaUIsR0FBRyxtQkFBWSxDQUFDLE9BQU8sQ0FBQyxlQUFlLENBQUMsTUFBTSxDQUFDLENBQUM7WUFDdkUsSUFBSSxpQkFBaUIsWUFBWSw0QkFBWSxFQUFFO2dCQUM3QyxNQUFNLFVBQVUsR0FBRyxpQkFBaUIsQ0FBQyxNQUFNLENBQUM7Z0JBQzVDLElBQUksVUFBVSxZQUFZLHdCQUFTLEVBQUU7b0JBQ25DLE1BQU0sR0FBRyxlQUFNLENBQUMsYUFBYSxDQUFDLFVBQVUsQ0FBQyxDQUFDO2lCQUMzQzthQUNGO1NBQ0Y7UUFDRCxJQUFJLENBQUMsTUFBTSxFQUFFO1lBQ1gsTUFBTSxHQUFHLGVBQU0sQ0FBQyxjQUFjLENBQUMsZUFBZSxFQUFFLGdCQUFnQixFQUFFLGVBQWUsQ0FBQyxNQUFNLENBQUMsQ0FBQztTQUMzRjtRQUVELE1BQU0sR0FBRyxHQUFHLElBQUksS0FBTSxTQUFRLFlBQVk7WUFBMUI7O2dCQUNFLGFBQVEsR0FBRyx3QkFBYyxDQUFDLFFBQVEsQ0FBQyxlQUFlLENBQUMsY0FBYyxDQUFDLENBQUM7WUFDckYsQ0FBQztTQUFBLENBQUMsZUFBZSxFQUFFLEVBQUUsRUFBRTtZQUNyQixNQUFNO1NBQ1AsQ0FBQyxDQUFDO1FBQ0gsd0NBQXdDO1FBQ3hDLE1BQU0sQ0FBQyxNQUFNLEdBQUcsR0FBRyxDQUFDO1FBQ3BCLE9BQU8sR0FBRyxDQUFDO0tBQ1o7SUE2QkQ7OztPQUdHO0lBQ0ksa0JBQWtCLENBQUMsYUFBNEI7UUFDcEQsSUFBSSxDQUFDLFFBQVEsQ0FBQyxrQkFBa0IsQ0FBQyxhQUFhLENBQUMsQ0FBQztLQUNqRDs7QUExRUgsb0NBMkVDIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0IHsgUG9saWN5RG9jdW1lbnQgfSBmcm9tICdAYXdzLWNkay9hd3MtaWFtJztcbmltcG9ydCB7IFJlbW92YWxQb2xpY3ksIFJlc291cmNlLCBUb2tlbiwgVG9rZW5pemF0aW9uIH0gZnJvbSAnQGF3cy1jZGsvY29yZSc7XG5pbXBvcnQgeyBDZm5SZWZlcmVuY2UgfSBmcm9tICdAYXdzLWNkay9jb3JlL2xpYi9wcml2YXRlL2Nmbi1yZWZlcmVuY2UnO1xuaW1wb3J0IHsgQ29uc3RydWN0IH0gZnJvbSAnY29uc3RydWN0cyc7XG5pbXBvcnQgeyBCdWNrZXQsIElCdWNrZXQgfSBmcm9tICcuL2J1Y2tldCc7XG5pbXBvcnQgeyBDZm5CdWNrZXQsIENmbkJ1Y2tldFBvbGljeSB9IGZyb20gJy4vczMuZ2VuZXJhdGVkJztcblxuZXhwb3J0IGludGVyZmFjZSBCdWNrZXRQb2xpY3lQcm9wcyB7XG4gIC8qKlxuICAgKiBUaGUgQW1hem9uIFMzIGJ1Y2tldCB0aGF0IHRoZSBwb2xpY3kgYXBwbGllcyB0by5cbiAgICovXG4gIHJlYWRvbmx5IGJ1Y2tldDogSUJ1Y2tldDtcblxuICAvKipcbiAgICogUG9saWN5IHRvIGFwcGx5IHdoZW4gdGhlIHBvbGljeSBpcyByZW1vdmVkIGZyb20gdGhpcyBzdGFjay5cbiAgICpcbiAgICogQGRlZmF1bHQgLSBSZW1vdmFsUG9saWN5LkRFU1RST1kuXG4gICAqL1xuICByZWFkb25seSByZW1vdmFsUG9saWN5PzogUmVtb3ZhbFBvbGljeTtcbn1cblxuLyoqXG4gKiBUaGUgYnVja2V0IHBvbGljeSBmb3IgYW4gQW1hem9uIFMzIGJ1Y2tldFxuICpcbiAqIFBvbGljaWVzIGRlZmluZSB0aGUgb3BlcmF0aW9ucyB0aGF0IGFyZSBhbGxvd2VkIG9uIHRoaXMgcmVzb3VyY2UuXG4gKlxuICogWW91IGFsbW9zdCBuZXZlciBuZWVkIHRvIGRlZmluZSB0aGlzIGNvbnN0cnVjdCBkaXJlY3RseS5cbiAqXG4gKiBBbGwgQVdTIHJlc291cmNlcyB0aGF0IHN1cHBvcnQgcmVzb3VyY2UgcG9saWNpZXMgaGF2ZSBhIG1ldGhvZCBjYWxsZWRcbiAqIGBhZGRUb1Jlc291cmNlUG9saWN5KClgLCB3aGljaCB3aWxsIGF1dG9tYXRpY2FsbHkgY3JlYXRlIGEgbmV3IHJlc291cmNlXG4gKiBwb2xpY3kgaWYgb25lIGRvZXNuJ3QgZXhpc3QgeWV0LCBvdGhlcndpc2UgaXQgd2lsbCBhZGQgdG8gdGhlIGV4aXN0aW5nXG4gKiBwb2xpY3kuXG4gKlxuICogUHJlZmVyIHRvIHVzZSBgYWRkVG9SZXNvdXJjZVBvbGljeSgpYCBpbnN0ZWFkLlxuICovXG5leHBvcnQgY2xhc3MgQnVja2V0UG9saWN5IGV4dGVuZHMgUmVzb3VyY2Uge1xuICAvKipcbiAgICogQ3JlYXRlIGEgbXV0YWJsZSBgQnVja2V0UG9saWN5YCBmcm9tIGEgYENmbkJ1Y2tldFBvbGljeWAuXG4gICAqL1xuICBwdWJsaWMgc3RhdGljIGZyb21DZm5CdWNrZXRQb2xpY3koY2ZuQnVja2V0UG9saWN5OiBDZm5CdWNrZXRQb2xpY3kpOiBCdWNrZXRQb2xpY3kge1xuICAgIC8vIHVzZSBhIFwid2VpcmRcIiBpZCB0aGF0IGhhcyBhIGhpZ2hlciBjaGFuY2Ugb2YgYmVpbmcgdW5pcXVlXG4gICAgY29uc3QgaWQgPSAnQEZyb21DZm5CdWNrZXRQb2xpY3knO1xuXG4gICAgLy8gaWYgZnJvbUNmbkJ1Y2tldFBvbGljeSgpIHdhcyBhbHJlYWR5IGNhbGxlZCBvbiB0aGlzIENmbkJ1Y2tldFBvbGljeSxcbiAgICAvLyByZXR1cm4gdGhlIHNhbWUgTDJcbiAgICAvLyAoYXMgZGlmZmVyZW50IEwycyB3b3VsZCBjb25mbGljdCwgYmVjYXVzZSBvZiB0aGUgbXV0YXRpb24gb2YgdGhlIGRvY3VtZW50IHByb3BlcnR5IG9mIHRoZSBMMSBiZWxvdylcbiAgICBjb25zdCBleGlzdGluZyA9IGNmbkJ1Y2tldFBvbGljeS5ub2RlLnRyeUZpbmRDaGlsZChpZCk7XG4gICAgaWYgKGV4aXN0aW5nKSB7XG4gICAgICByZXR1cm4gPEJ1Y2tldFBvbGljeT5leGlzdGluZztcbiAgICB9XG5cbiAgICAvLyByZXNvbHZlIHRoZSBCdWNrZXQgdGhpcyBQb2xpY3kgcmVmZXJlbmNlc1xuICAgIGxldCBidWNrZXQ6IElCdWNrZXQgfCB1bmRlZmluZWQ7XG4gICAgaWYgKFRva2VuLmlzVW5yZXNvbHZlZChjZm5CdWNrZXRQb2xpY3kuYnVja2V0KSkge1xuICAgICAgY29uc3QgYnVja2V0SVJlc29sdmFibGUgPSBUb2tlbml6YXRpb24ucmV2ZXJzZShjZm5CdWNrZXRQb2xpY3kuYnVja2V0KTtcbiAgICAgIGlmIChidWNrZXRJUmVzb2x2YWJsZSBpbnN0YW5jZW9mIENmblJlZmVyZW5jZSkge1xuICAgICAgICBjb25zdCBjZm5FbGVtZW50ID0gYnVja2V0SVJlc29sdmFibGUudGFyZ2V0O1xuICAgICAgICBpZiAoY2ZuRWxlbWVudCBpbnN0YW5jZW9mIENmbkJ1Y2tldCkge1xuICAgICAgICAgIGJ1Y2tldCA9IEJ1Y2tldC5mcm9tQ2ZuQnVja2V0KGNmbkVsZW1lbnQpO1xuICAgICAgICB9XG4gICAgICB9XG4gICAgfVxuICAgIGlmICghYnVja2V0KSB7XG4gICAgICBidWNrZXQgPSBCdWNrZXQuZnJvbUJ1Y2tldE5hbWUoY2ZuQnVja2V0UG9saWN5LCAnQEZyb21DZm5CdWNrZXQnLCBjZm5CdWNrZXRQb2xpY3kuYnVja2V0KTtcbiAgICB9XG5cbiAgICBjb25zdCByZXQgPSBuZXcgY2xhc3MgZXh0ZW5kcyBCdWNrZXRQb2xpY3kge1xuICAgICAgcHVibGljIHJlYWRvbmx5IGRvY3VtZW50ID0gUG9saWN5RG9jdW1lbnQuZnJvbUpzb24oY2ZuQnVja2V0UG9saWN5LnBvbGljeURvY3VtZW50KTtcbiAgICB9KGNmbkJ1Y2tldFBvbGljeSwgaWQsIHtcbiAgICAgIGJ1Y2tldCxcbiAgICB9KTtcbiAgICAvLyBtYXJrIHRoZSBCdWNrZXQgYXMgaGF2aW5nIHRoaXMgUG9saWN5XG4gICAgYnVja2V0LnBvbGljeSA9IHJldDtcbiAgICByZXR1cm4gcmV0O1xuICB9XG5cbiAgLyoqXG4gICAqIEEgcG9saWN5IGRvY3VtZW50IGNvbnRhaW5pbmcgcGVybWlzc2lvbnMgdG8gYWRkIHRvIHRoZSBzcGVjaWZpZWQgYnVja2V0LlxuICAgKiBGb3IgbW9yZSBpbmZvcm1hdGlvbiwgc2VlIEFjY2VzcyBQb2xpY3kgTGFuZ3VhZ2UgT3ZlcnZpZXcgaW4gdGhlIEFtYXpvblxuICAgKiBTaW1wbGUgU3RvcmFnZSBTZXJ2aWNlIERldmVsb3BlciBHdWlkZS5cbiAgICovXG4gIHB1YmxpYyByZWFkb25seSBkb2N1bWVudCA9IG5ldyBQb2xpY3lEb2N1bWVudCgpO1xuXG4gIC8qKiBUaGUgQnVja2V0IHRoaXMgUG9saWN5IGFwcGxpZXMgdG8uICovXG4gIHB1YmxpYyByZWFkb25seSBidWNrZXQ6IElCdWNrZXQ7XG5cbiAgcHJpdmF0ZSByZXNvdXJjZTogQ2ZuQnVja2V0UG9saWN5O1xuXG4gIGNvbnN0cnVjdG9yKHNjb3BlOiBDb25zdHJ1Y3QsIGlkOiBzdHJpbmcsIHByb3BzOiBCdWNrZXRQb2xpY3lQcm9wcykge1xuICAgIHN1cGVyKHNjb3BlLCBpZCk7XG5cbiAgICB0aGlzLmJ1Y2tldCA9IHByb3BzLmJ1Y2tldDtcblxuICAgIHRoaXMucmVzb3VyY2UgPSBuZXcgQ2ZuQnVja2V0UG9saWN5KHRoaXMsICdSZXNvdXJjZScsIHtcbiAgICAgIGJ1Y2tldDogdGhpcy5idWNrZXQuYnVja2V0TmFtZSxcbiAgICAgIHBvbGljeURvY3VtZW50OiB0aGlzLmRvY3VtZW50LFxuICAgIH0pO1xuXG4gICAgaWYgKHByb3BzLnJlbW92YWxQb2xpY3kpIHtcbiAgICAgIHRoaXMucmVzb3VyY2UuYXBwbHlSZW1vdmFsUG9saWN5KHByb3BzLnJlbW92YWxQb2xpY3kpO1xuICAgIH1cbiAgfVxuXG4gIC8qKlxuICAgKiBTZXRzIHRoZSByZW1vdmFsIHBvbGljeSBmb3IgdGhlIEJ1Y2tldFBvbGljeS5cbiAgICogQHBhcmFtIHJlbW92YWxQb2xpY3kgdGhlIFJlbW92YWxQb2xpY3kgdG8gc2V0LlxuICAgKi9cbiAgcHVibGljIGFwcGx5UmVtb3ZhbFBvbGljeShyZW1vdmFsUG9saWN5OiBSZW1vdmFsUG9saWN5KSB7XG4gICAgdGhpcy5yZXNvdXJjZS5hcHBseVJlbW92YWxQb2xpY3kocmVtb3ZhbFBvbGljeSk7XG4gIH1cbn1cbiJdfQ==