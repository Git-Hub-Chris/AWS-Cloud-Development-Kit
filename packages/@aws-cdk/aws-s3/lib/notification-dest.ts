import cdk = require('@aws-cdk/cdk');
import { Bucket } from './bucket';

/**
 * Implemented by constructs that can be used as bucket notification targets.
 */
export interface INotificationDestination {
    /**
     * Registers this resource to receive notifications for the specified bucket.
     * @param bucket The bucket. Use the `path` of the bucket as a unique ID.
     */
    bucketNotificationDestination(bucket: Bucket): NotificationDestinationProps;
}

/**
 * Represents the properties of a notification target.
 */
export interface NotificationDestinationProps {
    /**
     * The notification type.
     */
    readonly type: NotificationDestinationType;

    /**
     * The ARN of the target.
     */
    readonly arn: cdk.Arn;
}

/**
 * Supported types of notificaiton targets.
 */
export enum NotificationDestinationType {
    Lambda,
    Queue,
    Topic
}
