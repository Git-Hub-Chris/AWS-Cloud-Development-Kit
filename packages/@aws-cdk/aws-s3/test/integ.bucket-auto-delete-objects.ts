import { App, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';
import { Construct } from 'constructs';
import * as s3 from '../lib';

class TestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new s3.Bucket(this, 'Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
  }
}

const app = new App();
new TestStack(app, 'cdk-s3-bucket-auto-delete-objects');
app.synth();
