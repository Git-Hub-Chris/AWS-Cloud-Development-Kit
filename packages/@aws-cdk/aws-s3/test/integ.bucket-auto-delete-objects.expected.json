{
  "Resources": {
    "Bucket83908E77": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "BucketPolicyE9A3008A": {
      "Type": "AWS::S3::BucketPolicy",
      "Properties": {
        "Bucket": {
          "Ref": "Bucket83908E77"
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn"
                  ]
                }
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket83908E77",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket83908E77",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "BucketAutoDeleteObjectsCustomResourceBAFD23C2": {
      "Type": "Custom::S3AutoDeleteObjects",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn"
          ]
        },
        "BucketName": {
          "Ref": "Bucket83908E77"
        }
      },
      "DependsOn": [
        "BucketPolicyE9A3008A"
      ],
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ]
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
          }
        ]
      }
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316S3Bucket54B0BA6B"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316S3VersionKey5D05A836"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316S3VersionKey5D05A836"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Timeout": 900,
        "MemorySize": 128,
        "Handler": "__entrypoint__.handler",
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092"
      ]
    }
  },
  "Parameters": {
    "AssetParameters94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316S3Bucket54B0BA6B": {
      "Type": "String",
      "Description": "S3 bucket for asset \"94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316\""
    },
    "AssetParameters94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316S3VersionKey5D05A836": {
      "Type": "String",
      "Description": "S3 key for asset version \"94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316\""
    },
    "AssetParameters94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316ArtifactHashCBB49470": {
      "Type": "String",
      "Description": "Artifact hash for asset \"94e60b8e13d1f4ded38f4813a8053fbe137c982a0b3c929daa8166d33f844316\""
    }
  }
}